task_sdk/src/airflow/sdk/api/datamodels/ti.py

```diff
@@ -0,0 +1,33 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import uuid
+from typing import Optional
+
+import msgspec
+
+
+class TaskInstance(msgspec.Struct, omit_defaults=True):
+    id: uuid.UUID
+
+    task_id: str
+    dag_id: str
+    run_id: str
+    try_number: int
+    map_index: Optional[int] = None  # noqa: UP007 msgspec doesn't support pipe-op
```

Comment:
The main reason we have -1 in the server side code is that since map_index is part of the primary key it can't be nullable. Null is conceptually a better indicator of "I'm not mapped"

Reply:
Like that idea. Was always bothered with the -1 ... but now I see it was not a badly introduced legacy but primary key is a reason I did not think of so far :-D